src/Common/src/Interop/Windows/kernel32/Interop.GetConsoleTitle.cs

```diff
@@ -20,53 +21,69 @@ internal partial class Kernel32
-        // Instead of doing complicated and dangerous logic to determine if this may or may not occur,
-        // we simply assume the worst and reserve a bigger buffer. This way we may use a bit more memory,
-        // but we will always be safe.
-        private static int CharCountToByteCount(int numChars)
-        {
-            return numChars * 2;
-        }
-
         // 1. We first try to call the GetConsoleTitle with an InitialBufferSizeInChars of 256.
         // 2. Then based on the return value either increase the capacity or return failure.
         internal static int GetConsoleTitle(out string title, out int titleLength)
         {
             int lastError = 0;
-            StringBuilder sb = new StringBuilder(CharCountToByteCount(InitialBufferSizeInChars + 1));
-            int len = GetConsoleTitle(sb, sb.Capacity);
 
-            if (len <= 0)
-            {
-                lastError = Marshal.GetLastWin32Error();
-            }
-            else if (len > MaxAllowedBufferSizeInChars)
-            {
-                // Title is greater than the allowed buffer so we do not read the title and only pass the length to the caller.
-                title = string.Empty;
-                titleLength = len;
-                return 0;
-            }
-            else
+            // See note below for why we're pessimistically setting the capacity as if it was bytes, even though it's in chars.
+            StringBuilder sb = new StringBuilder(InitialBufferSizeInChars * sizeof(char));
+
+            while (true)
             {
-                if (len > InitialBufferSizeInChars)
+                // If capacity is insufficient, sometimes this function returns length,
+                // and sometimes (Windows 10 RS2) it returns 0 with either ERROR_INSUFFICIENT_BUFFER or ERROR_SUCCESS.
+                // In either of those latter cases, try again hopefully with a larger buffer.
+                //
+                // In some cases (Windows 7, perhaps depending on codepage) it interprets the second parameter as bytes. 
+                // Give it characters, pssimistically, but make sure that the number of characters we offer is twice what it claims it needs.
+                //
+                // In some cases (Windows 10 RS2), the returned length includes null, in others (Windows 7) it does not. 
+                // Pessimistically assume it does not.
+                int len = GetConsoleTitle(sb, sb.Capacity);
+
+                if (len <= 0)
                 {
-                    // We need to increase the sb capacity and retry.
-                    sb.Capacity = CharCountToByteCount(len + 1);
-                    len = GetConsoleTitle(sb, sb.Capacity);
-                    if (len <= 0)
+                    lastError = Marshal.GetLastWin32Error();
+
+                    if (len < 0 || (lastError != Errors.ERROR_INSUFFICIENT_BUFFER && lastError != Errors.ERROR_SUCCESS))
```

Comment:
The ERROR_SUCCESS worries me- that is typically an issue of not getting LastError right away- I'm skeptical of that coming from the API.  Additionally, this shouldn't be checking length again.